planner/core/runtime_filter.go

```diff
@@ -0,0 +1,257 @@
+// RuntimeFilter struct:
+//
+//	id: 0
+//	buildNode: HashJoin
+//	srcExprList: [t2.k1]
+//	targetExprList: [t1.k1]
+//	rfType:  IN
+//	rfMode: LOCAL
+//	buildNodeID: 2
+//	targetNodeID: 0
+//
+// Although srcExprList and targetExprList attributes are lists, in fact, there will only be one value in them at present.
+// The reason why it is designed as a list is because:
+// Reserve an interface for subsequent multiple join equivalent expressions corresponding to one rf.
+// (Since IN supports multiple columns, this construction can be more efficient at the execution level)
+type RuntimeFilter struct {
+	// runtime filter id, unique in one query plan
+	id             int
+	buildNode      *PhysicalHashJoin
+	srcExprList    []*expression.Column
+	targetExprList []*expression.Column
+	rfType         RuntimeFilterType
+	// The following properties need to be set after assigning a scan node to RF
+	rfMode     RuntimeFilterMode
+	targetNode *PhysicalTableScan
+	// The plan id will be set when runtime filter clone()
+	// It is only used for runtime filter pb
+	buildNodeID  int
+	targetNodeID int
+}
+
+// NewRuntimeFilter construct runtime filter by Join and equal predicate of Join
+func NewRuntimeFilter(rfIDGenerator *util.IDGenerator, eqPredicate *expression.ScalarFunction, buildNode *PhysicalHashJoin) ([]*RuntimeFilter, int64) {
+	rightSideIsBuild := buildNode.RightIsBuildSide()
+	var srcExprList []*expression.Column
+	var targetExprUniqueID int64
+	if rightSideIsBuild {
+		srcExprList = append(srcExprList, eqPredicate.GetArgs()[1].(*expression.Column))
+		targetExprUniqueID = eqPredicate.GetArgs()[0].(*expression.Column).UniqueID
+	} else {
+		srcExprList = append(srcExprList, eqPredicate.GetArgs()[0].(*expression.Column))
+		targetExprUniqueID = eqPredicate.GetArgs()[1].(*expression.Column).UniqueID
+	}
+
+	rfTypes := buildNode.ctx.GetSessionVars().GetRuntimeFilterTypes()
+	result := make([]*RuntimeFilter, 0, len(rfTypes))
+	for _, rfType := range rfTypes {
+		rf := &RuntimeFilter{
+			id:          rfIDGenerator.GetNextID(),
+			buildNode:   buildNode,
+			srcExprList: srcExprList,
+			rfType:      rfType,
+		}
+		result = append(result, rf)
+	}
+	return result, targetExprUniqueID
+}
+
+func (rf *RuntimeFilter) assign(targetNode *PhysicalTableScan, targetExpr *expression.Column) {
+	rf.targetNode = targetNode
+	if len(rf.targetNode.runtimeFilterList) == 0 {
+		// todo use session variables instead
+		rf.targetNode.maxWaitTimeMs = 10000
+	}
+	rf.targetExprList = append(rf.targetExprList, targetExpr)
+	rf.buildNode.runtimeFilterList = append(rf.buildNode.runtimeFilterList, rf)
+	rf.targetNode.runtimeFilterList = append(rf.targetNode.runtimeFilterList, rf)
+	logutil.BgLogger().Debug("Assign RF to target node",
+		zap.String("RuntimeFilter", rf.String()))
+}
+
+// ExplainInfo explain info of runtime filter
+func (rf *RuntimeFilter) ExplainInfo(isBuildNode bool) string {
+	var builder strings.Builder
+	fmt.Fprintf(&builder, "%d[%s]", rf.id, rf.rfType)
```

Comment:
But I have actually commented that he is a runtimefilter in general. If a node has multiple runtime filters it will look like:
runtime filter: 0[IN] xxx, 1[IN] xxx